File: sdk-py/smolvm/microvm.py
"""smolvm Python SDK — MicroVM.

High-level wrapper for persistent MicroVMs.
MicroVMs differ from Sandboxes: different REST schema, persistent by design,
no image/container management.
"""

from __future__ import annotations

from base64 import b64decode, b64encode

from .client import SmolvmHttpClient
from .types import ExecOptions, ExecResult, MicroVMInfo


class MicroVM:
    """Stateful wrapper around a single smolvm MicroVM."""

    def __init__(self, name: str, http: SmolvmHttpClient):
        self.name = name
        self._http = http
        self._info: MicroVMInfo | None = None

    @property
    def state(self) -> str:
        return self._info.state if self._info else "unknown"

    # --------------------------------------------------------------------------
    # Lifecycle
    # --------------------------------------------------------------------------

    async def start(self) -> None:
        self._info = await self._http.start_microvm(self.name)

    async def stop(self) -> None:
        self._info = await self._http.stop_microvm(self.name)

    async def delete(self, force: bool = False) -> None:
        await self._http.delete_microvm(self.name, force)
        self._info = None

    async def info(self) -> MicroVMInfo:
        self._info = await self._http.get_microvm(self.name)
        return self._info

    async def cleanup(self) -> None:
        """Stop + delete. Ignores errors (safe for cleanup)."""
        try:
            await self.stop()
        except Exception:
            pass
        try:
            await self.delete()
        except Exception:
            pass

    # --------------------------------------------------------------------------
    # Execution
    # --------------------------------------------------------------------------

    async def exec(
        self, command: list[str], opts: ExecOptions | None = None
    ) -> ExecResult:
        return await self._http.exec_microvm(self.name, command, opts)

    async def sh(
        self, cmd: str, opts: ExecOptions | None = None
    ) -> ExecResult:
        return await self.exec(["sh", "-c", cmd], opts)

    async def run_command(
        self, cmd: str, opts: ExecOptions | None = None
    ) -> ExecResult:
        return await self.sh(cmd, opts)

    # --------------------------------------------------------------------------
    # File I/O (via exec channel)
    # --------------------------------------------------------------------------

    async def write_file(self, path: str, content: str) -> None:
        encoded = b64encode(content.encode()).decode()
        last_slash = path.rfind("/")
        if last_slash > 0:
            await self.sh(f"mkdir -p '{path[:last_slash]}'")
        result = await self.sh(f"echo '{encoded}' | base64 -d > '{path}'")
        if result.exit_code != 0:
            raise RuntimeError(f"write_file failed: {result.stderr}")

    async def read_file(self, path: str) -> str:
        result = await self.sh(f"base64 '{path}'")
        if result.exit_code != 0:
            raise RuntimeError(f"read_file failed: {result.stderr}")
        return b64decode(result.stdout.strip()).decode()

    async def write_files(self, files: dict[str, str]) -> None:
        for path, content in files.items():
            await self.write_file(path, content)

File: sdk-ts/microvm.ts
/**
 * smolvm TypeScript SDK — MicroVM
 *
 * High-level wrapper for persistent MicroVMs.
 * MicroVMs differ from Machinees: different REST schema, persistent by design,
 * no image/container management.
 */

import { SmolvmHttpClient } from "./client.ts";
import type {
  ExecOptions,
  ExecResult,
  MicroVMInfo,
} from "./types.ts";

export class MicroVM {
  private _info: MicroVMInfo | null = null;

  constructor(
    public readonly name: string,
    private readonly http: SmolvmHttpClient,
  ) {}

  get state(): string {
    return this._info?.state ?? "unknown";
  }

  // --------------------------------------------------------------------------
  // Lifecycle
  // --------------------------------------------------------------------------

  async start(): Promise<void> {
    this._info = await this.http.startMicroVM(this.name);
  }

  async stop(): Promise<void> {
    this._info = await this.http.stopMicroVM(this.name);
  }

  async delete(force = false): Promise<void> {
    await this.http.deleteMicroVM(this.name, force);
    this._info = null;
  }

  async info(): Promise<MicroVMInfo> {
    this._info = await this.http.getMicroVM(this.name);
    return this._info;
  }

  async cleanup(): Promise<void> {
    try { await this.stop(); } catch { /* ignore */ }
    try { await this.delete(); } catch { /* ignore */ }
  }

  // --------------------------------------------------------------------------
  // Execution
  // --------------------------------------------------------------------------

  async exec(command: string[], opts?: ExecOptions): Promise<ExecResult> {
    return this.http.execMicroVM(this.name, command, opts);
  }

  async sh(cmd: string, opts?: ExecOptions): Promise<ExecResult> {
    return this.exec(["sh", "-c", cmd], opts);
  }

  async runCommand(cmd: string, opts?: ExecOptions): Promise<ExecResult> {
    return this.sh(cmd, opts);
  }

  // --------------------------------------------------------------------------
  // File I/O (same exec-channel approach as Machine)
  // --------------------------------------------------------------------------

  async writeFile(path: string, content: string): Promise<void> {
    const encoded = btoa(content);
    const dir = path.substring(0, path.lastIndexOf("/"));
    if (dir) {
      await this.sh(`mkdir -p '${dir}'`);
    }
    const result = await this.sh(`echo '${encoded}' | base64 -d > '${path}'`);
    if (result.exitCode !== 0) {
      throw new Error(`writeFile failed: ${result.stderr}`);
    }
  }

  async readFile(path: string): Promise<string> {
    const result = await this.sh(`base64 '${path}'`);
    if (result.exitCode !== 0) {
      throw new Error(`readFile failed: ${result.stderr}`);
    }
    return atob(result.stdout.trim());
  }

  async writeFiles(files: Record<string, string>): Promise<void> {
    for (const [path, content] of Object.entries(files)) {
      await this.writeFile(path, content);
    }
  }
}
